repose-aggregator/commons/utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletResponseWrapper.scala

```diff
@@ -63,67 +76,275 @@ class HttpServletResponseWrapper(originalResponse: HttpServletResponse, headerMo
+   * Formats the input time (as milliseconds since the epoch) to a format defined in RFC2616.
    */
-  override def appendHeader(headerName: String, headerValue: String, quality: Double): Unit = ???
+  override def addDateHeader(name: String, timeSinceEpoch: Long): Unit =
+    addHeader(name, DateUtils.formatDate(new Date(timeSinceEpoch)))
 
   /**
    * @throws IllegalStateException when headerMode is anything other than ResponseMode.MUTABLE
    */
-  override def removeHeader(headerName: String): Unit = ???
+  override def appendHeader(name: String, value: String): Unit = {
+    ifMutable(headerMode) {
+      val existingValues = getHeaders(name)
+      existingValues.lastOption match {
+        case Some(currentLastValue) =>
+          val newLastValue = currentLastValue + "," + value
+          headerMap = headerMap + (name -> (existingValues.dropRight(1).toSeq :+ newLastValue))
+        case None => addHeader(name, value)
+      }
+    }
+  }
 
   /**
    * @throws IllegalStateException when headerMode is anything other than ResponseMode.MUTABLE
    */
-  override def replaceHeader(headerName: String, headerValue: String): Unit = ???
+  override def appendHeader(name: String, value: String, quality: Double): Unit =
+    appendHeader(name, s"$value;q=$quality")
+
+  override def setHeader(name: String, value: String): Unit = {
+    headerMap = headerMap + (name -> Seq(value))
+
+    if (headerMode != ResponseMode.MUTABLE) {
+      // Write through to the wrapped response immediately
+      super.setHeader(name, value)
+    }
+  }
+
+  override def setIntHeader(name: String, value: Int): Unit = setHeader(name, value.toString)
+
+  override def setDateHeader(name: String, timeSinceEpoch: Long): Unit =
+    setHeader(name, DateUtils.formatDate(new Date(timeSinceEpoch)))
 
   /**
    * @throws IllegalStateException when headerMode is anything other than ResponseMode.MUTABLE
    */
-  override def replaceHeader(headerName: String, headerValue: String, quality: Double): Unit = ???
+  override def removeHeader(name: String): Unit = {
+    ifMutable(headerMode) {
+      headerMap = headerMap - name
+    }
+  }
+
+  override def replaceHeader(name: String, value: String): Unit = setHeader(name, value)
+
+  override def replaceHeader(name: String, value: String, quality: Double): Unit = setHeader(name, s"$value;q=$quality")
+
+  override def setContentLength(i: Int): Unit = {
+    if (!isCommitted) {
+      setIntHeader(CommonHttpHeader.CONTENT_LENGTH.toString, i)
+    }
+  }
 
-  override def getHeadersList(headerName: String): util.List[String] = ???
+  override def getContentType: String = getHeader(CommonHttpHeader.CONTENT_TYPE.toString)
 
-  override def getPreferredHeaders(headerName: String): util.List[String] = ???
+  override def setContentType(s: String): Unit = {
+    if (!isCommitted) {
+      val contentType =
+        if (responseBodyType != ResponseBodyType.PrintWriter) s
```

Comment:
I should probably name that method parameter.

This just sets the content type to what it should be. That is, it either strips the character encoding or leaves it alone. According to the docs on ServletRequest, the character encoding should not be set if getWriter has been called.

http://docs.oracle.com/javaee/6/api/javax/servlet/ServletResponse.html#setContentType(java.lang.String)